pkg/meshdiscovery/config/appmesh/syncer.go

```
@@ -0,0 +1,111 @@
+	emitter := v1.NewAppmeshDiscoveryEmitter(
+		cl.cs.Discovery.Mesh,
+		cl.cs.Input.Pod,
+		cl.cs.Input.Upstream,
+	)
+	reconciler := v1.NewMeshReconciler(cl.cs.Discovery.Mesh)
+	syncer := newAppmeshDiscoveryConfigSyncer(reconciler)
+	el := v1.NewAppmeshDiscoveryEventLoop(emitter, syncer)
+
+	return el, nil
+}
+
+func newAppmeshDiscoveryConfigSyncer(reconciler v1.MeshReconciler) *appmeshDiscoveryConfigSyncer {
+	return &appmeshDiscoveryConfigSyncer{reconciler: reconciler}
+}
+
+type appmeshDiscoveryConfigSyncer struct {
+	reconciler v1.MeshReconciler
+}
+
+func (s *appmeshDiscoveryConfigSyncer) Sync(ctx context.Context, snap *v1.AppmeshDiscoverySnapshot) error {
+	ctx = contextutils.WithLogger(ctx, fmt.Sprintf("appmesh-config-discovery-sync-%v", snap.Hash()))
+	logger := contextutils.LoggerFrom(ctx)
+	pods, meshes, upstreams := snap.Pods.List(), snap.Meshes.List(), snap.Upstreams.List()
+	fields := []interface{}{
+		zap.Int("meshes", len(meshes)),
+		zap.Int("pods", len(pods)),
+		zap.Int("upstreams", len(upstreams)),
+	}
+
+	logger.Infow("begin sync", fields...)
+	defer logger.Infow("end sync", fields...)
+	logger.Debugf("full snapshot: %v", snap)
+
+	appmeshMeshes := utils.GetMeshes(snap.Meshes.List(), utils.AppmeshFilterFunc, utils.FilterByLabels(appmesh.DiscoverySelector))
+
+	var updatedMeshes v1.MeshList
+	for _, mesh := range appmeshMeshes {
+		config, err := appmeshcfg.NewAwsAppMeshConfiguration(mesh.Metadata.Name, pods, upstreams)
+		if err != nil {
+			return err
+		}
+		if typedConfig, ok := config.(*appmeshcfg.AwsAppMeshConfigurationImpl); ok {
+			updatedMeshes = append(updatedMeshes, updatedMesh(mesh, typedConfig))
+		}
+	}
+
+	listOpts := clients.ListOpts{
+		Ctx:      ctx,
+		Selector: appmesh.DiscoverySelector,
+	}
+	return s.reconciler.Reconcile("", updatedMeshes, nil, listOpts)
+}
+
+func updatedMesh(mesh *v1.Mesh, config *appmeshcfg.AwsAppMeshConfigurationImpl) *v1.Mesh {
+	result := mesh
+	awsMesh := mesh.GetAwsAppMesh()
+	if awsMesh == nil {
```

Since this should never happen (we pass in a list of app meshes), I would return an error here to notify of the unexpected behavior.